kv/memberlist/node_zone_aware_routing.go

```diff
@@ -96,18 +98,82 @@ func newZoneAwareNodeSelectionDelegate(localRole NodeRole, localZone string, log
 	}
 }
 
-// SelectNode implements memberlist.NodeSelectionDelegate.
-// It determines whether a remote node should be selected for gossip operations and whether it should be preferred.
-func (d *zoneAwareNodeSelectionDelegate) SelectNode(node memberlist.Node) (selected, preferred bool) {
-	// Decode remote node metadata.
-	remoteMeta := EncodedNodeMetadata(node.Meta)
-	remoteZone := remoteMeta.Zone()
-	remoteRole := remoteMeta.Role()
+// SelectNodes implements memberlist.NodeSelectionDelegate.
+// It determines which remote nodes should be selected for gossip operations and which one should be preferred.
+func (d *zoneAwareNodeSelectionDelegate) SelectNodes(nodes []*memberlist.NodeState) (selected []*memberlist.NodeState, preferred *memberlist.NodeState) {
+	if d.localRole != NodeRoleMember && d.localRole != NodeRoleBridge {
+		level.Warn(d.logger).Log("msg", "memberlist zone-aware routing is running with an unknown role", "role", d.localRole)
+	}
+
+	// Skip zone-aware routing if local zone is not set.
+	if d.localZone == "" {
+		return nodes, nil
+	}
+
+	// Pre-allocate backing arrays on the stack for up to 5 zones (common case).
+	// Slices will grow on the heap only if more zones are encountered.
+	var zonesWithMembersBacking [5]string
+	zonesWithMembers := zonesWithMembersBacking[:0]
+	var zonesWithAliveBridgesBacking [5]string
+	zonesWithAliveBridges := zonesWithAliveBridgesBacking[:0]
```

Comment:
I was wondering why the compiler wouldn't optimise this... and I had to check, and turns out this stuff we do is unnecessary. [The generated code is the same](https://godbolt.org/z/zE8Thvozj)

```suggestion
	// Pre-allocate backing arrays on the stack for up to 5 zones (common case).
	zonesWithMembers := make([]string, 0, 5)
	zonesWithAliveBridges := make([]string, 0, 5)
```